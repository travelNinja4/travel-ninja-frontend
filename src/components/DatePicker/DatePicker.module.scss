/**
------------------------------------------------------
styles specific to the DatePicker component
------------------------------------------------------
**/

.datePickerWrapper {
  position: relative;
  display: inline-block;
  // width: 100%;
  // padding: 0.8rem;
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.08);
  transition: all 0.2s ease-in-out;
  border-radius: 8px;
  background-color: #fff;
  border: 1px solid #ddd;

  &:hover {
    box-shadow: 0 6px 14px rgba(0, 0, 0, 0.1);
  }
}

.calendarIcon {
  position: absolute;
  right: 0.75rem;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
  color: #000000;
  transition: transform 0.2s ease;
}

/* Month-Year header */
// .header {
//   display: flex;
//   justify-content: center;
//   gap: 0.5rem;
//   margin-bottom: 0.8rem;
// }

/* DayPicker Base Styles */
.dayPicker {
  font-family: Arial, sans-serif;
  color: #000000;
}

/* Weekday header */
.dayPicker .rdp-weekday {
  font-weight: 600;
  color: #555;
}

/* Day cells */
.dayPicker .rdp-day {
  border-radius: 4px;
  padding: 8px;
  text-align: center;
}

/* Selected range */
.dayPicker .rdp-day_selected,
.dayPicker .rdp-day_range_start,
.dayPicker .rdp-day_range_end {
  background-color: #007bff;
  color: #fff;
}

/* Range in between */
.dayPicker .rdp-day_range_middle {
  background-color: #cce5ff;
}

/* Disabled days */
.dayPicker .rdp-day_disabled {
  color: #ccc;
  pointer-events: none;
}

/* Navigation buttons */
.dayPicker .rdp-nav_button {
  background: none;
  border: none;
  color: #007bff;
  font-weight: bold;
  cursor: pointer;
}

/* ---------------------------------------------------
   Inline Mode Styles (dropdown below text field)
--------------------------------------------------- */
.inlineContainer {
  position: absolute;
  top: 100%;
  left: 0;
  margin-top: 0.5rem;
  z-index: 100;
  background: #fff;
  border: 1px solid #ddd;
  border-radius: 8px;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  animation: fadeIn 0.2s ease-in-out;
}

/* ---------------------------------------------------
   Modal Mode Styles (centered overlay)
--------------------------------------------------- */
.modalOverlay {
  position: fixed;
  inset: 0;
  background: rgba(0, 0, 0, 0.4);
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 1000;
  animation: fadeIn 0.2s ease-in-out;
}

.modalContainer {
  background: #fff;
  padding: 20px;
  border-radius: 12px;
  box-shadow: 0 4px 20px rgba(0, 0, 0, 0.2);
  width: 90%;
  max-width: 360px;
  animation: scaleIn 0.25s ease-in-out;
}

/* ---------------------------------------------------
   Animations
--------------------------------------------------- */
@keyframes fadeIn {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

@keyframes scaleIn {
  from {
    opacity: 0;
    transform: scale(0.95);
  }
  to {
    opacity: 1;
    transform: scale(1);
  }
}
