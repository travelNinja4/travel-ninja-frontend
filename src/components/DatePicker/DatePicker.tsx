/**
 * A customizable date picker that integrates easily with forms and supports themes, disabled dates, and validation.
 *
 * @example
 * ```tsx
 * import DatePicker from '@src/components/DatePicker'
 *
 * export default function DatePicker() {
 *   return <DatePicker label="Hello" />;
 * }
 * ```
 */
'use client';

import { DayPicker, DateRange } from 'react-day-picker';
import 'react-day-picker/style.css';
import styles from './DatePicker.module.scss';
import { useMemo, useState } from 'react';
import TextField from '../TextField';
import { Calendar } from 'lucide-react';

/**
 * Define the props available for the DatePicker component.
 */
interface DatePickerProps {
  label?: string;
  /**
   * Where to show the calendar popup:
   * - "inline" = below the text field
   * - "modal" = centered overlay
   */
  displayMode?: 'inline' | 'modal';

  /**
   * The mode prop determines the selection mode.
   * When the mode prop is set to "single", only one date can be selected at a time
   * Set the mode prop to "multiple" to enable the selection of multiple dates in DayPicker.
   * Set the mode prop to "range" to enable the selection of a continuous range of dates in DayPicker.
   */
  mode?: 'single' | 'multiple' | 'range';

  /**
   *
   */
  min?: number;

  /**
   *
   */
  max?: number;

  initialRange?: DateRange;
  navLayout?: 'around' | 'after' | undefined;

  /**
   * To prevent the user from navigating between months, set the disableNavigation prop to true.
   */
  disableNavigation?: boolean;
  disabled?: Date[];
  inputClassName?: string;
  onChange?: (value: Date | Date[] | DateRange | undefined) => void;
}

export default function DatePicker({
  displayMode = 'inline',
  label,
  mode = 'single',
  initialRange,
  navLayout = 'around',
  disableNavigation = false,
  min,
  max,
  disabled,
  inputClassName,
  onChange,
}: DatePickerProps) {
  const [open, setOpen] = useState(false);
  const [month, setMonth] = useState(new Date());
  const [selected, setSelected] = useState<Date>();
  const [selectedMultiple, setSelectedMultiple] = useState<Date[]>();
  const [range, setRange] = useState<DateRange | undefined>(initialRange);

  const handleSelect = (value: any) => {
    if (mode === 'range') setRange(value);
    else if (mode === 'single') setSelected(value);
    else if (mode === 'multiple') setSelectedMultiple(value);

    onChange?.(value);
    setOpen(false);
  };

  // Helper to render DayPicker for all modes
  const renderPicker = () => {
    const props = {
      month,
      onMonthChange: setMonth,
      navLayout,
      disableNavigation,
      pagedNavigation: true,
      fixedWeeks: true,
      disabled,
      className: styles.dayPicker,
      onSelect: handleSelect,
    };

    if (mode === 'single')
      return (
        <DayPicker
          {...props}
          mode="single"
          captionLayout="dropdown"
          reverseYears
          reverseMonths
          startMonth={new Date(2024, 6)}
          endMonth={new Date(2025, 9)}
          selected={selected}
        />
      );
    if (mode === 'multiple')
      return <DayPicker {...props} mode="multiple" selected={selectedMultiple} />;
    return (
      <DayPicker
        {...props}
        mode="range"
        selected={range}
        min={min}
        max={max}
        captionLayout="dropdown"
        reverseYears
        reverseMonths
        startMonth={new Date(2024, 6)}
        endMonth={new Date(2025, 9)}
      />
    );
  };

  return (
    <div data-testid="DatePickerTest">
      <TextField
        readOnly
        label={label}
        placeholder="Select date"
        onClick={() => !disabled && setOpen((prev) => !prev)}
        className={inputClassName}
        isCalender
      />

      {open && (
        <div className={styles.datePickerWrapper}>
          <>
            {displayMode === 'inline' ? (
              <div className={styles.inlineContainer}>{renderPicker()}</div>
            ) : (
              <div className={styles.modalOverlay} onClick={() => setOpen(false)}>
                <div className={styles.modalContainer} onClick={(e) => e.stopPropagation()}>
                  {renderPicker()}
                </div>
              </div>
            )}
          </>
        </div>
      )}
    </div>
  );
}
