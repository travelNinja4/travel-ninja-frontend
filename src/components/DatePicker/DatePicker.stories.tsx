import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import DatePicker from './DatePicker'

const meta = {
  title: 'components/DatePicker',
  component: DatePicker,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: { 
        control: 'text', 
        description: 'Sample label for the component' 
        },
  },
  args: { 
    /** onClick: fn() **/
    },
} satisfies Meta<typeof DatePicker>;

export default meta;
type Story = StoryObj<typeof meta>;