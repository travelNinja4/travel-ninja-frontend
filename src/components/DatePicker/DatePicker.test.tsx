import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import DatePicker from './DatePicker'

describe('Home', () => {
  /** Base props for DatePicker **/
  const baseProps={
    label: 'DatePicker Test'
  }
 
  /**
  * @test
  * Renders the DatePicker component without throwing an error
  * and checks if the DatePickerTest text is present in the document.
  *
  * This ensures that the component loads correctly with the base props.
  */
  it('renders without crashing', () => {
   render(<DatePicker {...baseProps}/>);
   expect(screen.getByTestId('DatePickerTest')).toBeInTheDocument();
  });

  /**
  * @test
  * Ensures that the DatePicker component renders without throwing any runtime errors.
  *
  * This is a safety check to confirm that the component can mount successfully
  * with the provided base props, even if it does not render any visible content.
  */
  it('renders without throwing', () => {
  expect(() => render(<DatePicker {...baseProps}/>)).not.toThrow();
  });

  /**
  * @test
  * Captures a snapshot of the DatePicker component with base props
  * and ensures it matches the previously stored snapshot.
  *
  * This helps detect any unexpected UI changes.
  */
  it('matches snapshot', () => {
   const { asFragment } = render(<DatePicker {...baseProps}/>);
   expect(asFragment()).toMatchSnapshot();
  });
})