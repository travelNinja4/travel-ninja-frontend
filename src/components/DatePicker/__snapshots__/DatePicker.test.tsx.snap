// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="TextFieldTest"
  >
    <label
      class="typography label align-left label"
      data-testid="TypographyTest"
    >
      DatePicker Test 
    </label>
    <div
      class="inputWrapper"
    >
      <input
        class="inputField"
        placeholder="Select date"
        readonly=""
        type="text"
      />
      <div
        class="iconButton"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-calendar"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 2v4"
          />
          <path
            d="M16 2v4"
          />
          <rect
            height="18"
            rx="2"
            width="18"
            x="3"
            y="4"
          />
          <path
            d="M3 10h18"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
